### PR TITLE
#34 Actually parse duration-like fields

### LINKED ISSUE #34: Parse episode duration value
Currently, we treat `Episode.ITunes#duration` as raw strings. We should accept them only if they are in one of the following formats:

* `[hours]:[minutes]:[seconds]` (max two digits per hh/mm/ss block)
* `[minutes]:[seconds]` (max two digits per mm/ss block)
* `[total_seconds]`

See: https://support.google.com/podcast-publishers/answer/9889544

They should be stored as `Duration`, like we already do for `podcast:soundbite`. GIven there are multiple possible ways to represent the values, it'd be a good idea to wrap the `Duration` in a sealed class that represents the style in which the value is serialised (e.g., `SecondsDuration`, `HhMmDuration`, `HhMmSsDuration`) — we should use it for the soundbite duration, too, although that can only be expressed as seconds.

### PR DESCRIPTION
**Pull request by rock3r**
This PR closes #34 by implementing a proper duration values parsing logic. We store those values internally as `java.time.Duration`, but we wrap them with sealed classes to make sure that we persist the original styling.

There's a ton of complications in parsing, and formatting, the durations, mostly due to the counterintuitive ways Duration works (this surely had me banging my head on my desk!). Thankfully tests have guided me through the darkness :)

Now the iTunes duration is expressed by a `StyledDuration?`, and the PodcastIndex soundbites start and duration are expressed by a `SecondsAndFraction`. To note, there is _some_ loss of idempotency here due to some cleaning up of the values we get when parsing and writing them (e.g., a duration of `00:10` will become `0:10`, and a duration of `1.120000` will become `1.12`). To maintain full transparency we'd have to make this mechanism a lot more complicated, and I don't think that's what we want to do. The values themselves aren't altered, and that's where I'd draw the line.

### CODE REVIEW

**Review by mpgirro — CHANGES_REQUESTED** [review]
Nice work! Durations prove to be a quite a pain. Hope your head is alright 🙂 

There are already quite some hoops here to jump through, so the loss of idempotency is fine in my opinion, since there is just a loss of characters, but no loss of temporal information.

File: src/main/kotlin/dev/stalla/model/StyledDuration.kt [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        @JvmOverloads
+        public fun secondsAndFraction(seconds: Int = 0, nano: Long = 0, positive: Boolean = true): SecondsAndFraction {
+            require(seconds >= 0 && nano >= 0) { "Components cannot be negative" }
+            val sign = if (positive) "" else "-"
+            // This is not ideal but Duration does not handle negative nano-only inputs well otherwise
+            val durationString = "PT$sign$seconds.${nano.toString().padStart(9, '0')}S"
+            return SecondsAndFraction(Duration.parse(durationString))
+        }
+
+        public fun seconds(seconds: Int = 0, positive: Boolean = true): Seconds {
```
[the remark this answers is not in the document]
Comment (rock3r) [reply src/main/kotlin/dev/stalla/model/StyledDuration.kt:170]: Yes! Forgot :)

File: src/main/kotlin/dev/stalla/model/StyledDuration.kt [hunk 2]
```diff
@@ -0,0 +1,197 @@
+package dev.stalla.model
+
+import dev.stalla.util.InternalApi
+import java.time.Duration
+import kotlin.math.absoluteValue
+import kotlin.math.pow
+
+public sealed class StyledDuration {
```
[the remark this answers is not in the document]
Comment (rock3r) [reply src/main/kotlin/dev/stalla/model/StyledDuration.kt:8]: Sure, I meant to then by the time it was all working it was late and I forgot 😅

[the remark this answers is not in the document]
Comment (mpgirro) [reply src/main/kotlin/dev/stalla/model/StyledDuration.kt:8]: Don't worry, I know these moments just all to well^^

File: src/main/kotlin/dev/stalla/model/StyledDuration.kt [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+    public companion object Factory {
+
+        // Parses the format: [-][HH:]MM:SS
+        private val hhMmSsRegex = "(-)?(?:(\\d{1,2}):)?(\\d{1,2}):(\\d{1,2})".toRegex()
+
+        // Parses the format: [-]SS...[.FFF...]
+        private val plainSecondsRegex = "(-)?(\\d+)(?:\\.(\\d+))?".toRegex()
+
+        public fun of(rawDuration: String?): StyledDuration? {
```
[the remark this answers is not in the document]
Comment (rock3r) [reply src/main/kotlin/dev/stalla/model/StyledDuration.kt:113]: Maybe `rawValue`?

File: src/test/java/dev/stalla/model/EpisodeBuilderFactoryTest.java [hunk 4]
```diff
@@ -117,8 +115,8 @@ public void testEpisodeBitloveBuilderFactory() {
     @DisplayName("should build an Episode.Podcast model using builder factory methods only")
     public void testEpisodePodcastBuilderFactory() {
         EpisodePodcastindexSoundbiteBuilder episodePodcastindexSoundbiteBuilder = Soundbite.builder()
-            .startTime(Duration.ZERO)
-            .duration(Duration.ZERO);
+            .startTime(StyledDuration.Factory.secondsAndFraction(0))
```
[the remark this answers is not in the document]
Comment (mpgirro) [reply src/test/java/dev/stalla/model/EpisodeBuilderFactoryTest.java:118]: Actually I just noticed that the other `of` methods don't have a `@JvmStatic` either. Ups 😅

File: src/main/kotlin/dev/stalla/model/StyledDuration.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (mpgirro) [reply src/main/kotlin/dev/stalla/model/StyledDuration.kt]: Yeah, sounds good!

[the remark this answers is not in the document]
Comment (rock3r) [reply src/main/kotlin/dev/stalla/model/StyledDuration.kt]: Ok, let me change that

File: src/test/java/dev/stalla/model/EpisodeBuilderFactoryTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (rock3r) [reply src/test/java/dev/stalla/model/EpisodeBuilderFactoryTest.java:119]: Yup fixed those too while I was at it :)

[the remark this answers is not in the document]
Comment (mpgirro) [reply src/test/java/dev/stalla/model/EpisodeBuilderFactoryTest.java:119]: Perfect 👍

**Comment by rock3r** [comment]
And all done

**Comment by mpgirro** [comment]
LGTM 👍

**Pull request merged**